src/devices/wifi/nm-iwd-manager.c

```diff
@@ -845,6 +896,9 @@ nm_iwd_manager_init (NMIwdManager *self)
 	g_signal_connect (priv->manager, NM_MANAGER_DEVICE_ADDED,
 	                  G_CALLBACK (device_added), self);
 
+	g_signal_connect (NM_SETTINGS_GET, NM_SETTINGS_SIGNAL_CONNECTION_REMOVED,
```

Comment:
So my `g_signal_handlers_disconnect_by_data (NM_SETTINGS_GET, self);` is specifically in a destructor for the singleton.  For simplicity I could even completely skip this line although it still seems correct, so I think the second simplest thing is for NMIwdManager to also keep a reference to the settings singleton.

On the other hand I note that nm_settings_get () is actually defined in nm-manager.c and uses the NMManager's refcounted pointer to the settings singleton, and as you note we already keep a reference to the NMManager.

Reply:
Well, currently most singleton instances are leaked anyway, and none of this can happen. However, that's a bug, that should be fixed and requires that such issues are handled correctly today.

I don't think you could omit disconnecting the handlers, because if NMSettings emits the signal after iwd-manager is gone, would lead to crash. Assuming that NMSettings wouldn't/shouldn't do that, seems fragile.

Yes, if you look at the implementation of `nm_settings_get()`, then `NMSettings` lives at least as long as `NMManager`, and you keep the manager alive. Well, sort of. `dispose()` may be invoked multiple times (theoretically, probably not in practice in NetworkManager), so you would need at least check `if (manager) { g_signal_handlers_disconnect_by_data (NM_SETTINGS_GET, self); }`.

Better let's just not make so many assumptions, and keep a reference to the settings singleton as well. It seems to me a good rule of thumb: if you subscribe signal handlers, you also need to keep a reference to the source.